text_volpiano_alignment.py

```diff
@@ -0,0 +1,196 @@
+Database (more details about these conventions can be found 
+at https://cantusdatabase.org/documents). See the README for more information.
+
+Use align_text_and_volpiano to align the text and melody strings of a chant.
+"""
+import re
+import logging
+from itertools import zip_longest, takewhile
+from cantus_text_syllabification import syllabify_text
+
+# A string containing all valid volpiano characters in Cantus
+# Database. Used to check for invalid characters.
+VALID_VOLPIANO_CHARS = "-19abcdefghijklmnopqrsyz)ABCDEFGHIJKLMNOPQRSYZ3467]"
+
+
+def _preprocess_volpiano(volpiano_str: str) -> str:
+    """
+    Prepares volpiano string for alignment with text:
+    - Checks for any invalid characters
+    - Ensures proper spacing around barlines ("3" and "4") and missing
+        music indicators ("6")
+    - Ensures volpiano string has an ending barline ("3" or "4")
+
+    volpiano_str [str]: volpiano string
+
+    returns [str]: preprocessed volpiano string
+    """
+    processed_str = ""
+    volpiano_str_len = len(volpiano_str)
+    for i, char in enumerate(volpiano_str):
+        # Check if char is valid
+        if char not in VALID_VOLPIANO_CHARS:
+            logging.debug("Removed invalid character (%s) in volpiano string.", char)
+            continue
+        # Check if char is a barline or missing music indicator and ensure
+        # proper spacing
+        if (char in "346") and (i != volpiano_str_len - 1):
+            # Add proper spacing before barline
+            while processed_str[-3:] != "---":
+                processed_str += "-"
+            # Add barline character
+            processed_str += char
+            # Add proper spacing after barline
+            num_hyph_ahead = sum(
+                1 for _ in takewhile(lambda x: x == "-", volpiano_str[i + 1 :])
+            )
+            processed_str += "-" * (3 - num_hyph_ahead)
+            continue
+        processed_str += char
+    # Ensure volpiano string ends with a properly-spaced barline
+    if processed_str[-4:] not in ["---3", "---4"]:
+        processed_str = processed_str.rstrip("-") + "---3"
+    logging.debug("Preprocessed volpiano string: %s", processed_str)
+    return processed_str
+
+
+def _postprocess_spacing(
+    comb_text_and_vol: "list[tuple[str, str]]",
+) -> "list[tuple[str, str]]":
+    """
+    Handle some special spacing requirements for optimal display
+    of volpiano with chant text.
+
+    Ensures that:
+     - the length of missing music sections responds to the length of
+        the text associated with the section
+
+    comb_text_and_vol [list[tuple[str, str]]]:
+        list of tuples of text syllables and volpiano syllables
+    """
+    comb_text_and_vol_rev_spacing = []
+    for text_elem, vol_elem in comb_text_and_vol:
+        if vol_elem[0] == "6":
+            text_length = len(text_elem)
+            if text_length > 10:
+                vol_elem = "6---" + "---" * (text_length // 3) + "6---"
```

Comment:
This function is good! Just curious: why not `vol_elem = "6---" + "-" * text_length + "6---"`? Why are we doing it in groups of 3?